### PR TITLE
Wav2Vec2 CTC in Flax/JAX

### PR DESCRIPTION
**Pull request by patrickvonplaten**
Adds code for Wav2Vec2 CTC training:

# TODO
- [x] Make sure loss is equal -> run `check_flax_ctc.py`
- [x] Make sure training scripts runs correctly
- [x] Reproduce a librispeech 100h training run from PyTorch

### CODE REVIEW

File: models/modeling_flax_wav2vec2.py [hunk 1]
```diff
@@ -682,6 +682,7 @@ def __init__(
         input_shape: Tuple = (1, 1024),
         seed: int = 0,
         dtype: jnp.dtype = jnp.float32,
+        _do_init: bool = True,
```
Comment (patrickvonplaten) [inline models/modeling_flax_wav2vec2.py:685]: Need this for newer versions of `transformers`

File: run_flax_speech_recognition_ctc.py [hunk 2]
```diff
@@ -0,0 +1,1219 @@
...
+            params=new_params,
+            opt_state=to_dtype(new_opt_state),
+            **kwargs,
+        )
+
+    @classmethod
+    def create(cls, *, apply_fn, params, tx, to_dtype, **kwargs):
+        """Creates a new instance with `step=0` and initialized `opt_state`."""
+        # downcast optimizer state to bf16 if mixed-precision training
+        opt_state = tx.init(to_dtype(params)) if tx is not None else None
```
Comment (sanchit-gandhi) [inline run_flax_speech_recognition_ctc.py:341]: Nice! Clean way of keeping the TrainState during eval only!

File: run_flax_speech_recognition_ctc.py [hunk 3]
```diff
@@ -0,0 +1,1219 @@
...
+            grad, loss = jax.lax.scan(accum_minibatch_step, init_grad, batch)
+
+        # update state
+        new_state = state.apply_gradients(
+            grads=grad,
+            dropout_rng=new_dropout_rng,
+            to_dtype=to_dtype,
+        )
+
+        # compute gradient norms over all layers, total encoder, total decoder and global for detailed monitoring
```
Comment (sanchit-gandhi) [inline run_flax_speech_recognition_ctc.py:1069]:
```suggestion
        # compute gradient norms over all layers and globally for detailed monitoring
```

File: run_flax_speech_recognition_ctc.py [hunk 4]
```diff
@@ -0,0 +1,1219 @@
...
+        layer_grad_norm = jax.tree_map(jnp.linalg.norm, grad)
+        logs = {
+            "layer_grad_norm": layer_grad_norm,
+            "grad_norm": jnp.linalg.norm(jax.tree_util.tree_leaves(layer_grad_norm)),
+        }
+
+        # compute parameter norms over all layers, total encoder, total decoder and global for detailed monitoring
```
Comment (sanchit-gandhi) [inline run_flax_speech_recognition_ctc.py:1076]:
```suggestion
        # compute parameter norms over all layers and globally for detailed monitoring
```

**Review by sanchit-gandhi — APPROVED** [review]
The CTC model and training script look great! Thanks for adding this! I took a deep look into the CTC loss function - there doesn't seem to be anything to suggest why the eval loss was different during the training run and then the subsequent standalone evaluation ([here](https://wandb.ai/patrickvonplaten/flax-speech-recognition-ctc/runs/2otixvsm) and [here](https://wandb.ai/patrickvonplaten/flax-speech-recognition-ctc/runs/3ocravol)). Will perform a separate set of training + eval runs and see if we observe the same phenomenon.

File: run_flax_speech_recognition_ctc.py [hunk 5]
```diff
@@ -0,0 +1,1219 @@
...
+
+            # normalize eval metrics
+            eval_metrics = get_metrics(eval_metrics)
+            eval_metrics = jax.tree_map(jnp.mean, eval_metrics)
+            eval_metrics = to_fp32(eval_metrics)
+
+            # compute WER metric and get predicted string (for debugging)
+            wer_desc = ""
+            pred_str = []
+            label_str = []
```
Comment (sanchit-gandhi) [inline run_flax_speech_recognition_ctc.py:1191]: Can remove these as we always run compute metrics
```suggestion
```

File: run_flax_speech_recognition_ctc.py [hunk 6]
```diff
@@ -0,0 +1,1219 @@
...
+    tokenizer.save_pretrained(training_args.output_dir)
+    config.save_pretrained(training_args.output_dir)
+
+    processor = AutoProcessor.from_pretrained(training_args.output_dir)
+
+    data_collator = FlaxDataCollatorSpeechSeq2SeqWithPadding(
+        processor=processor,
+        input_padding="longest",
+        pad_input_to_multiple_of=pad_input_to_multiple_of,
+        max_label_length=512,  # TODO(PVP, Sanchit) - to check here
```
[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply run_flax_speech_recognition_ctc.py:903]: Cool! Think it should be good though. The following math should work quite well here:

max logit length = max input_length / 320 (downsample factor wav2vec2)

max label length should always be at most twice the max logit length

=> max input length = 20s = 320,000
320,000 / 320 / 2 = 500

[the remark this answers is not in the document]
Comment (patrickvonplaten) [reply run_flax_speech_recognition_ctc.py:903]: What def worth checking with label length here!
